### PR TITLE
ch4: Fix persistent receive with MPI_ANY_SOURCE

### PR DESCRIPTION
**Pull request by hajimefu**
This PR fixes a bug in CH4 that it doesn't handle persistent receive with MPI_ANY_SOURCE correctly.

Mainly, there were two issues in the current CH4 implementation:

* In case of persistent receive, MPID_Irecv is not supposed to handle both shmmod and netmod (that is already done at the persistent request level).
* In case of ANY_SOURCE receive, MPID_Startall must consider a case where MPIDI_NM_mpi_startall completes right after the invocation (e.g. receive has matched an unexpected message). In this case, CH4 has to cancel SHM receive right away, otherwise nobody would mark the SHM request (hence the persistent request itself) complete. This fix was based on the idea in the Active Message Shared Memory PR (https://github.com/pmodels/mpich/pull/3199/), particularly the patch "ch4: Fix any_source receives" by Wesley Bland and Sergey Gubanov.

This is a continuation work of #3383.

### CODE REVIEW

File: src/mpid/ch4/src/ch4_types.h [hunk 1]
```diff
@@ -110,11 +110,12 @@ enum {
     MPIDI_CH4U_EPOTYPE_LOCK_ALL = 6       /**< MPI_Win_lock_all access epoch */
 };
 
-/* Enum for calling types between netmod and shm */
+/* Enum (bit mask fields) for calling types between netmod and shm */
```
[the remark this answers is not in the document]
Comment (hajimefu) [reply src/mpid/ch4/src/ch4_types.h]: Okay, so you are proposing to add such a macro, right? (I couldn't find `MPL_BIT` in the current code base)

[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpid/ch4/src/ch4_types.h]: Correct. We do not have it today.

[the remark this answers is not in the document]
Comment (hajimefu) [reply src/mpid/ch4/src/ch4_types.h]: Done

**Comment by hajimefu** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/most

**Comment by hajimefu** [comment]
test:jenkins/ch4/most

**Comment by hajimefu** [comment]
Rebased with the latest master (after the CH4 namespace cleanup)

File: src/mpid/ch4/src/ch4_types.h [hunk 2]
```diff
@@ -110,11 +110,12 @@ enum {
...
 enum {
-    MPIDI_NETMOD = 0,
-    MPIDI_SHM = 1
+    MPIDI_NETMOD = MPL_BIT(0),
+    MPIDI_SHM = MPL_BIT(1),
 };
+#define MPIDI_CH4R_TRANSPORT_ALL (MPIDI_NETMOD | MPIDI_SHM)
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4_types.h]: Just noticed an old naming convention. Will fix.

Comment (hajimefu) [reply src/mpid/ch4/src/ch4_types.h]: Done.

**Comment by hajimefu** [comment]
test:jenkins/ch4/most

**Comment by hajimefu** [comment]
Rebased onto the latest master.

**Comment by hajimefu** [comment]
test:jenkins/ch4/most

File: src/mpid/ch4/src/ch4_recv.h [hunk 3]
```diff
@@ -288,21 +306,24 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_irecv_safe(void *buf,
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IRECV_SAFE);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IRECV_SAFE);
 
-    MPID_THREAD_SAFE_BEGIN(VCI, MPIDI_global.vci_lock, cs_acq);
+    MPIR_Assert((transport & (MPIDI_NETMOD | MPIDI_SHM)) != 0);
+
+    MPID_THREAD_SAFE_BEGIN(VCI, MPIDI_Global.vci_lock, cs_acq);
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4_recv.h]: oops -- will fix.

**Comment by hajimefu** [comment]
test:jenkins/ch4/most
test:jenkins/ch4/vci

**Comment by raffenet** [comment]
Fixed conflicts. Re-testing. I signed off on some of the patches, but I need to more time to digest the larger ones.
test:jenkins/ch4/most

File: src/mpid/ch4/src/ch4_recv.h [hunk 4]
```diff
@@ -250,9 +266,10 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_recv_safe(void *buf,
         *(req) = MPIR_Request_create(MPIR_REQUEST_KIND__RECV);
         MPIR_ERR_CHKANDSTMT((*req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
         MPIR_Datatype_add_ref_if_not_builtin(datatype);
-        MPIDI_workq_pt2pt_enqueue(RECV, NULL /*send_buf */ , buf, count, datatype,
-                                  rank, tag, comm, context_offset, av,
-                                  status, *req, NULL /*flag */ , NULL /*message */ ,
+        MPIDI_workq_pt2pt_enqueue(RECV, 0 /* transport */ , NULL /*send_buf */ , buf, count,
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_recv.h]: Use enum value here for transport argument?

File: src/mpid/ch4/src/ch4_recv.h [hunk 5]
```diff
@@ -332,8 +353,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_imrecv_safe(void *buf,
         MPIR_Request *request = MPIR_Request_create(MPIR_REQUEST_KIND__RECV);
         MPIR_ERR_CHKANDSTMT(request == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
         MPIR_Datatype_add_ref_if_not_builtin(datatype);
-        MPIDI_workq_pt2pt_enqueue(IMRECV, NULL /*send_buf */ , buf, count, datatype,
-                                  0 /*rank */ , 0 /*tag */ , NULL /*comm */ ,
+        MPIDI_workq_pt2pt_enqueue(IMRECV, 0 /* transport */ , NULL /*send_buf */ , buf, count,
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_recv.h]: Same here and a few more places in this commit.

File: src/mpid/ch4/src/ch4_recv.h (hunk 4 above)
Comment (hajimefu) [reply src/mpid/ch4/src/ch4_recv.h]: Done

File: src/mpid/ch4/src/ch4_recv.h (hunk 5 above)
Comment (hajimefu) [reply src/mpid/ch4/src/ch4_recv.h]: Done

**Comment by hajimefu** [comment]
test:jenkins/ch4/most

**Comment by hajimefu** [comment]
@raffenet Thanks for rebasing and reviewing this. I think I addressed your comments. Could you take another look?

**Comment by raffenet** [comment]
I don't love the API modifications, but I don't have any good alternative off the top of my head. @halimamer could you take a look at this and share any thoughts? If this looks OK to you, I will merge it.

**Comment by raffenet** [comment]
OK, I have one thought off the top of my head. Would it be simpler to modify `recv_init` to not pre-compute anything and therefore treat these exactly like regular `MPI_ANY_SOURCE` operations? There's obviously some optimization lost if we do so, but it might be worth avoiding the added complexity for now.

**Comment by hajimefu** [comment]
Hmm, that's an interesting idea. Let me think about it.

With a quick thought, it'd deprecate some NM/SHM functions like `MPIDI_NM/SHM_mpi_recv_init`. We may also want to deprecate `NM/SHM_mpi_send_init` functions as well for consistency. I wouldn't worry too much about it and actually feel happy about reducing the interfaces. Just wanted to quickly check it this aligns with your thoughts.

**Comment by raffenet** [comment]
Just so we're on the same page, are we going to move ahead with my proposed solution? Should we close this PR or will you update it with the new code when it's ready?

**Comment by hajimefu** [comment]
@raffenet My plan was to update this PR, but if you want to close it for now, that's fine too.

**Comment by raffenet** [comment]
Re-using this PR is fine. I'll leave it open.

**Comment by hajimefu** [comment]
@raffenet I created a new PR (#3914) based on the idea discussed here. I'll close this PR.

**Pull request closed without merging**